docs/man/memcat.1
.TH "MEMCAT" "1" "September 28, 2011" "0.53" "libmemcached"
.SH NAME
memcat \- libmemcached Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" Man page generated from reStructeredText.
.
.SH SYNOPSIS
.sp
memcat [options] key
.sp
Copy a set of keys to stdout
.SH DESCRIPTION
.sp
\fBmemcat\fP outputs to stdout the value a single or multiple set of keys
stored in a memcached(1) server. If any key is not found an error is returned.
.sp
It is similar to the standard UNIX cat(1) utility.
.SH OPTIONS
.sp
You can specify servers via the option:
.INDENT 0.0
.TP
.B \-\-servers
.UNINDENT
.sp
or via the environment variable:
.INDENT 0.0
.TP
.B \(gaMEMCACHED_SERVERS\(ga
.UNINDENT
.sp
For a full list of operations run the tool with the option:
.INDENT 0.0
.TP
.B \-\-help
.UNINDENT
.SH HOME
.sp
To find out more information please check:
\fI\%http://libmemcached.org/\fP
.SH AUTHOR
.sp
Brian Aker, <\fI\%brian@tangent.org\fP>
.sp
Mark Atwood <\fI\%mark@fallenpegasus.com\fP>
.SH SEE ALSO
.sp
\fImemcached(1)\fP \fIlibmemcached(3)\fP
.SH AUTHOR
Brian Aker
.SH COPYRIGHT
2011, Brian Aker DataDifferential, http://datadifferential.com/
.\" Generated by docutils manpage writer.
.\"
.


ChangeLog
2024-09-02 [version 1.14.0]
## 🐣 New Features
* Add `mop upsert` method. #292
* Add multiple storage method.
  * `memcached_mset()`
  * `memcached_madd()`
  * `memcached_mreplace()`
  * `memcached_mprepend()`
  * `memcached_mpppend()`
  * `memcached_mcas()`
## 🐛 Bug Fixes
* Fix buffer size when calling `snprintf()` with leading whitespace and integer value. #304
* Use `__linux__` instead of `TARGET_OS_LINUX` to check if the current OS is Linux. #306
* Use different lock filenames for multiple proxy in a process
## 📝 Documentation
* Fix typo.
## ✅ Testing
* Cancel previous CI in progress when new commits are added. #296

2024-02-22 [version 1.13.3]
## 🔧 Enhancements
* Fix compile warnings on RHEL8 and MacOS.
* Do not use `#pragma once`.
* Select smallest host as owner when ketama hash collided. #237
* Add parameter validation in `arcus_proxy_connect()`.
* Move io reset from where calls `memcached_vdo()` to where fails inside of `memcached_vdo()`. #250
* Use `<latest-release>-unknown` version when git describe fails.
* Add a limit of count about bop mget/smget commands.
* Import `scanner.l` and `parser.yy` from libmemcached-0.53.
* Refactor ketama weighted related implementation. #263 #273 #274 #275 #276 #277
* Fix package name and info to 'arcus-c-client'.
## 🐛 Bug Fixes
* Call `arcus_server_check_for_update()` in `memcached_exist_by_key()`. #241
* Check `mc->connect_timeout` instead of `mc->poll_timeout` in `connect_poll()`. #247
* Remove a duplicate init macro in configure.ac. #267
* Return `hashkit_get_distribution_function()` instead of `hashkit_get_function()` in `memcached_behavior_get_distribution_hash()`.
## 📝 Documentation
* Add unit test FAQ.
* Fix inconsistencies between code and documentation.
## ✅ Testing
* Use Github Actions instead of Travis for CI test.

2022-02-22 [version 1.13.2]
        * [FEATURE] supported the 3 replicas in replication.
        * [FIX] return switchover error code when coll get with delete.
        * [FIX] fixed wrong aggregate pipe return code in case of failure.
        * [FIX] hold the pool lock when update the cache list of master mc.
        * [FIX] fixed the cache list update of member mc in proxy mode.
        * [ENHANCE] enhanced the cachelist update logic with serverinfo.
        * [ENHANCE] Not to close and reopen pool connections
        * [ENHANCE] removed the dual pool repopulation when start at first.
        * [ENHANCE] optimized the slave comparision in memcached_rgroup_update().
        * [ENHANCE] optimized the release processing in memcached_pool.
        * [ENHANCE] ensure more concurrency in memcached pool.
        * [CLEANUP] fixed the compile warnings.
        * [CLEANUP] refactored serveral parts of code.

2021-07-06 [version 1.13.1]
        * [FEATURE] enable zookeeper multi thread mode.
        * [FIX] if MEMCACHED_ERRNO entered in memcached_set_error(),
                return instead of assert.

2021-01-18 [version 1.13.0]
        * [FEATURE] supported mgets command.
        * [FIX] check space_sepatated_keys immediately after memcached_connect().
        * [FIX] check mget command support immediately after memcached_connect().
        * [INTERNAL] return MEMCACHED_CONNECTION_FAILURE in memcached_fetch_result().

2020-12-09 [version 1.12.0]
        * [FEATURE] set the max key size to 4000.
        * [DOC] changed max element bytes: 4KB => 16KB.
                (committed at 20200317 and included in 1.10.4)
        * [DOC] fixed the smget result API and structure names.

2020-12-07 [version 1.11.0]
        * [FEATURE] supported space-separated keys.
        * [FEATURE] supported mget command, efficient for getting multiple items.
        * [FEATURE] requested version operation when connect to server.
        * [FEATURE] increased max key length: 250 => 4K.
        * [ENHANCE] do memcached_quit_server() if poll timeout occurs in io_flush.
        * [ENHANCE] increased cluster max size for multi key op: 200 => 1000.
        * [IMPROVE] reduced string copy when convert eflag/update_filter to string.
        * [FIX] set purge recursion flag to false before return.
        * [FIX] handled the host failures in the multi key operation.
        * [FIX] unset noreply before get operation in noreply_test().
        * [FIX] fixed unallocated memory access in mkey fetch.
        * [OTHERS] Refactored some code for readabiblity.

2020-10-22 [version 1.10.5]
        * [FEATURE] bop incr/decr with initial option.
        * [FIX] registered "NULL" if hostname lookup fails in do_add_client_info().
        * [FIX] added the TYPE_MISMATCH handling in kv response
        * [DOC] changed the description of the max item size
        * [DOC] refactored the document format and some content.

2020-08-04 [version 1.10.4]
        * [FEATURE] Added memcached_detail_error_message() API.
        * [FEATURE] Immediate reconnect according to error responses.
        * [FEATURE] Changed max element bytes. 4KB => 16KB.
        * [FEATURE] Returned E2BIG if CLIENT_ERROR object too large.
        * [FEATURE] Changed MEMCACHED_SERVER_FAILURE_RETRY_TIMEOUT. 2sec => 1sec.
        * [ENHANCE] Enlarged the contents of error messages in detail.
        * [ENHANCE] Stored detail error message in error_messages field.
        * [ENHANCE] Added arcus ZK manager for processing ZK events.
        * [FIX] pipe operation is stopped if switchover is done.
        * [FIX] Fixed version check condition.
        * [FIX] Fixed several bugs in handling response strings.
        * [FIX] Added the key test where needed.
        * [FIX] Do client_info and watcher registration only when needed.
        * [FIX] Prevent master_mc from being freed while using pool.
        * [FIX] Changed to create version using m4 file.
        * [FIX] Fixed simple null pointer dereferencing.
        * [DOC] Added multi eflag filter description.
        * [Test] Added test case for PIPE_ERROR bad error.
        * [OTHERS] Refactored the response module for handling responses.

clients/memcat.cc
/* LibMemcached
 * Copyright (C) 2006-2009 Brian Aker
 * All rights reserved.
 *
 * Use and distribution licensed under the BSD license.  See
 * the COPYING file in the parent directory for full text.
 *
 * Summary:
 *
 */

#include <config.h>

#include <cstdio>
#include <cstring>
#include <getopt.h>
#include <iostream>
#include <unistd.h>
#include <libmemcached/memcached.h>

#include "utilities.h"

#define PROGRAM_NAME "memcat"
#define PROGRAM_DESCRIPTION "Cat a set of key values to stdout."


/* Prototypes */
void options_parse(int argc, char *argv[]);

static int opt_binary= 0;
static int opt_verbose= 0;
static int opt_displayflag= 0;
static char *opt_servers= NULL;
static char *opt_hash= NULL;
static char *opt_username;
static char *opt_passwd;
static char *opt_file;

int main(int argc, char *argv[])
{
  memcached_st *memc;
  char *string;
  size_t string_length;
  uint32_t flags;
  memcached_return_t rc;
  memcached_server_st *servers;

  int return_code= 0;

  options_parse(argc, argv);
  initialize_sockets();

  if (!opt_servers)
  {
    char *temp;

    if ((temp= getenv("MEMCACHED_SERVERS")))
      opt_servers= strdup(temp);
    else
    {
      fprintf(stderr, "No Servers provided\n");
      exit(1);
    }
  }

  memc= memcached_create(NULL);
  process_hash_option(memc, opt_hash);

  servers= memcached_servers_parse(opt_servers);

  memcached_server_push(memc, servers);
  memcached_server_list_free(servers);
  memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_BINARY_PROTOCOL,
                         (uint64_t)opt_binary);

  if (opt_username and LIBMEMCACHED_WITH_SASL_SUPPORT == 0)
  {
    memcached_free(memc);
    std::cerr << "--username was supplied, but binary was not built with SASL support." << std::endl;
    return EXIT_FAILURE;
  }

  if (opt_username)
  {
    memcached_return_t ret;
    if (memcached_failed(ret= memcached_set_sasl_auth_data(memc, opt_username, opt_passwd)))
    {
      std::cerr << memcached_last_error_message(memc) << std::endl;
      memcached_free(memc);
      return EXIT_FAILURE;
    }
  }

  while (optind < argc)
  {
    string= memcached_get(memc, argv[optind], strlen(argv[optind]),
                          &string_length, &flags, &rc);
    if (rc == MEMCACHED_SUCCESS)
    {
      if (opt_displayflag)
      {
        if (opt_verbose)
          printf("key: %s\nflags: ", argv[optind]);
        printf("%x\n", flags);
      }
      else
      {
        if (opt_verbose)
        {
          printf("key: %s\nflags: %x\nlength: %lu\nvalue: ",
                 argv[optind], flags, (unsigned long)string_length);
        }

        if (opt_file)
        {
          FILE *fp;
          size_t written;

          fp= fopen(opt_file, "w");
          if (!fp)
          {
            perror("fopen");
            return_code= -1;
            break;
          }

          written= fwrite(string, 1, string_length, fp);
          if (written != string_length)
          {
            fprintf(stderr, "error writing file (written %lu, should be %lu)\n", (unsigned long)written, (unsigned long)string_length);
            return_code= -1;
            break;
          }

          if (fclose(fp))
          {
            fprintf(stderr, "error closing file\n");
            return_code= -1;
            break;
          }
        }
        else
        {
            printf("%.*s\n", (int)string_length, string);
        }
        free(string);
      }
    }
    else if (rc != MEMCACHED_NOTFOUND)
    {
      fprintf(stderr, "memcat: %s: memcache error %s",
              argv[optind], memcached_strerror(memc, rc));
      if (memcached_last_error_errno(memc))
      {
        fprintf(stderr, " system error %s", strerror(memcached_last_error_errno(memc)));
      }
      fprintf(stderr, "\n");

      return_code= -1;
      break;
    }
    else // Unknown Issue
    {
      fprintf(stderr, "memcat: %s not found\n", argv[optind]);
      return_code= -1;
    }
    optind++;
  }

  memcached_free(memc);

  if (opt_servers)
    free(opt_servers);
  if (opt_hash)
    free(opt_hash);

  return return_code;
}


void options_parse(int argc, char *argv[])
{
  int option_index= 0;
  int option_rv;

  memcached_programs_help_st help_options[]=
  {
    {0},
  };

  static struct option long_options[]=
    {
      {(OPTIONSTRING)"version", no_argument, NULL, OPT_VERSION},
      {(OPTIONSTRING)"help", no_argument, NULL, OPT_HELP},
      {(OPTIONSTRING)"quiet", no_argument, NULL, OPT_QUIET},
      {(OPTIONSTRING)"verbose", no_argument, &opt_verbose, OPT_VERBOSE},
      {(OPTIONSTRING)"debug", no_argument, &opt_verbose, OPT_DEBUG},
      {(OPTIONSTRING)"servers", required_argument, NULL, OPT_SERVERS},
      {(OPTIONSTRING)"flag", no_argument, &opt_displayflag, OPT_FLAG},
      {(OPTIONSTRING)"hash", required_argument, NULL, OPT_HASH},
      {(OPTIONSTRING)"binary", no_argument, NULL, OPT_BINARY},
      {(OPTIONSTRING)"username", required_argument, NULL, OPT_USERNAME},
      {(OPTIONSTRING)"password", required_argument, NULL, OPT_PASSWD},
      {(OPTIONSTRING)"file", required_argument, NULL, OPT_FILE},
      {0, 0, 0, 0},
    };

  while (1)
  {
    option_rv= getopt_long(argc, argv, "Vhvds:", long_options, &option_index);
    if (option_rv == -1) break;
    switch (option_rv)
    {
    case 0:
      break;
    case OPT_BINARY:
      opt_binary = 1;
      break;
    case OPT_VERBOSE: /* --verbose or -v */
      opt_verbose = OPT_VERBOSE;
      break;
    case OPT_DEBUG: /* --debug or -d */
      opt_verbose = OPT_DEBUG;
      break;
    case OPT_VERSION: /* --version or -V */
      version_command(PROGRAM_NAME);
      break;
    case OPT_HELP: /* --help or -h */
      help_command(PROGRAM_NAME, PROGRAM_DESCRIPTION, long_options, help_options);
      break;
    case OPT_SERVERS: /* --servers or -s */
      opt_servers= strdup(optarg);
      break;
    case OPT_HASH:
      opt_hash= strdup(optarg);
      break;
    case OPT_USERNAME:
      opt_username= optarg;
      break;
    case OPT_PASSWD:
      opt_passwd= optarg;
      break;
    case OPT_FILE:
      opt_file= optarg;
      break;

    case OPT_QUIET:
      close_stdio();
      break;

    case '?':
      /* getopt_long already printed an error message. */
      exit(1);
    default:
      abort();
    }
  }
}
